# World University Rankings
## Data exploration
This notebook is intended to explore the data to see what fields are available, what values are allowed etc. We will use pandas library to read a CSV format and leverage built-in functions to investigate the data. 

We will use to data to answer some questions like: What are the top-10 Computer Science Universities in the world?

In [1]:
import pandas as pd

Pandas uses a data format called "dataframe". Think of it as a table with rows and columns. Let us open one file and see what the dataframe would look like. For the purpose of this exercise, we will use World University Rankings dataset available on Kaggle: https://www.kaggle.com/datasets/r1chardson/the-world-university-rankings-2011-2023

We are looking at data from 2022. Please use other years' data as needed.

In [2]:
df = pd.read_csv('../dataset/2022_rankings.csv')

In [3]:
df.shape

(2112 (0x840), 24 (0x18))

We see that the data has 2112 rows and 24 columns. Let us print the first 5 rows (along with the columns). This will give us an idea about what the data looks like.

In [4]:
df.head(5)

rank_order rank                                name scores_overall  \
0          10    1                University of Oxford           95.7   
1          20   =2  California Institute of Technology           95.0   
2          30   =2                  Harvard University           95.0   
3          40    4                 Stanford University           94.9   
4          50   =5             University of Cambridge           94.6   

   scores_overall_rank  scores_teaching  scores_teaching_rank  \
0                   10             91.0                     5   
1                   20             93.6                     2   
2                   30             94.5                     1   
3                   40             92.3                     3   
4                   50             90.9                     6   

   scores_research  scores_research_rank  scores_citations  ...  \
0             99.6                     1              98.0  ...   
1             96.9                     4              97.8  ...   
2             98.9                     3              99.2  ...   
3             96.8                     5              99.9  ...   
4             99.5                     2              96.2  ...   

   scores_international_outlook_rank        location  stats_number_students  \
0                                 26  United Kingdom                 20,835   
1                                167   United States                  2,233   
2                                209   United States                 21,574   
3                                211   United States                 16,319   
4                                 32  United Kingdom                 19,680   

   stats_student_staff_ratio  stats_pc_intl_students stats_female_male_ratio  \
0                       10.7                     42%                 47 : 53   
1                        6.3                     34%                 36 : 64   
2                        9.5                     24%                 50 : 50   
3                        7.3                     23%                 46 : 54   
4                       11.1                     39%                 47 : 53   

                                      aliases  \
0                        University of Oxford   
1  California Institute of Technology caltech   
2                          Harvard University   
3                         Stanford University   
4                     University of Cambridge   

                                    subjects_offered closed unaccredited  
0  Accounting & Finance,General Engineering,Commu...  False        False  
1  Languages, Literature & Linguistics,Economics ...  False        False  
2  Mathematics & Statistics,Civil Engineering,Lan...  False        False  
3  Physics & Astronomy,Computer Science,Politics ...  False        False  
4  Business & Management,General Engineering,Art,...  False        False  

[5 rows x 24 columns]

To get the number of rows and columns, use the __shape__ function.

In [5]:
rows = df.shape[0]
cols = df.shape[1]

In [6]:
print(rows)

2112


In [7]:
print(cols)

24


## Columns in the dataset
Let us also take a look at what columns exist in the dataset.

In [8]:
df.columns

Index(['rank_order', 'rank', 'name', 'scores_overall', 'scores_overall_rank',
       'scores_teaching', 'scores_teaching_rank', 'scores_research',
       'scores_research_rank', 'scores_citations', 'scores_citations_rank',
       'scores_industry_income', 'scores_industry_income_rank',
       'scores_international_outlook', 'scores_international_outlook_rank',
       'location', 'stats_number_students', 'stats_student_staff_ratio',
       'stats_pc_intl_students', 'stats_female_male_ratio', 'aliases',
       'subjects_offered', 'closed', 'unaccredited'],
      dtype='object')

## Extracting interesting data
Often, you will have a huge dataset but you will be interested in only a subset. It is helpful to understand how you can extract interesting/relevant data to reduce the size of the dataset that you need to work with.

For this exercise, let us focus on these columns only:
 - rank_order
 - rank
 - name
 - location
 - subjects_offered

And let us only look at the top-200 universities based on the rank.

To get top-N rows sorted by a column (rank_order), we will use this function:
DataFrame.sort_values as described here: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html


In our case, we need to pass the following args:
 - by = 'rank_order'
 - axis = 0 (because we want to get top-N rows)

We will also use "Slicing" in Python to get the first-N values. For example, to get the first 10 elements in an array/list, you will use: arr[:10]

In [9]:
top_200_universities = df.sort_values(by='rank_order', axis=0)[:200]

In [10]:
top_200_universities.shape

(200 (0xc8), 24 (0x18))

As you see above, we now have a dataframe which has only 200 rows. 

Now let us extract specific columns. To do this, you will need to create a list of columns that you want to keep. Then, you will use that list to filter the original dataframe to get the columns you need. 

For example, if you only need *col2* and *col6* from a dataframe, you will use this:

df = df[['col2', 'col6']]

In [11]:
cols_to_keep = ['rank_order', 'rank', 'name', 'location', 'subjects_offered']

In [12]:
# Create a copy so that the original dataset remains unchanged
df_2022_rankings = top_200_universities[cols_to_keep].copy()

In [13]:
df_2022_rankings

rank_order  rank                                name        location  \
0            10     1                University of Oxford  United Kingdom   
1            20    =2  California Institute of Technology   United States   
2            30    =2                  Harvard University   United States   
3            40     4                 Stanford University   United States   
4            50    =5             University of Cambridge  United Kingdom   
..          ...   ...                                 ...             ...   
195        1960   196          Medical University of Graz         Austria   
196        1970  =197    University of Erlangen-Nuremberg         Germany   
197        1980  =197                University of Geneva     Switzerland   
198        1990  =197             University of Hohenheim         Germany   
199        2000  =197         Sapienza University of Rome           Italy   

                                      subjects_offered  
0    Accounting & Finance,General Engineering,Commu...  
1    Languages, Literature & Linguistics,Economics ...  
2    Mathematics & Statistics,Civil Engineering,Lan...  
3    Physics & Astronomy,Computer Science,Politics ...  
4    Business & Management,General Engineering,Art,...  
..                                                 ...  
195                  Medicine & Dentistry,Other Health  
196  Archaeology,Computer Science,Sport Science,Bio...  
197  Politics & International Studies (incl Develop...  
198  Business & Management,Economics & Econometrics...  
199  Politics & International Studies (incl Develop...  

[200 rows x 5 columns]

As you can see above, you are left with 200 rows and 5 columns that are of interest.

## Making data-driven decisions
Now that you have a subset of the data, you can use that to answer questions that help you make decisions. Let us try to answer some questions like:
 - What are the top-10 Computer Science Universities?
 - What are the top-10 Computer Science Universities in the US?

To check whether "Computer Science" is offered in a university, we will use __str.contains__ function for a column.

This function will return a Boolean Series containing True or False depending on whether the pattern was found or not.
See documentation here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.contains.html


We will also add another column __is_computer_science__ that will specify whether or not Computer Science is offered for that university.

In [14]:
df_2022_rankings['is_computer_science'] = df_2022_rankings['subjects_offered'].str.contains('Computer Science')

In [15]:
df_2022_rankings

rank_order  rank                                name        location  \
0            10     1                University of Oxford  United Kingdom   
1            20    =2  California Institute of Technology   United States   
2            30    =2                  Harvard University   United States   
3            40     4                 Stanford University   United States   
4            50    =5             University of Cambridge  United Kingdom   
..          ...   ...                                 ...             ...   
195        1960   196          Medical University of Graz         Austria   
196        1970  =197    University of Erlangen-Nuremberg         Germany   
197        1980  =197                University of Geneva     Switzerland   
198        1990  =197             University of Hohenheim         Germany   
199        2000  =197         Sapienza University of Rome           Italy   

                                      subjects_offered  is_computer_science  
0    Accounting & Finance,General Engineering,Commu...                 True  
1    Languages, Literature & Linguistics,Economics ...                 True  
2    Mathematics & Statistics,Civil Engineering,Lan...                 True  
3    Physics & Astronomy,Computer Science,Politics ...                 True  
4    Business & Management,General Engineering,Art,...                 True  
..                                                 ...                  ...  
195                  Medicine & Dentistry,Other Health                False  
196  Archaeology,Computer Science,Sport Science,Bio...                 True  
197  Politics & International Studies (incl Develop...                 True  
198  Business & Management,Economics & Econometrics...                False  
199  Politics & International Studies (incl Develop...                 True  

[200 rows x 6 columns]

As you can see above, a new column __is_computer_science__ is created which contains True or False depending on whether the course is offered or not.

### 1. Top-10 Computer Science Universities
We will use the same "Slicing" as before to get the top-10 universities. Before that, we want to also get the universities that DO offer Computer Science. We will use the __is_computer_science__ column from before and get only those rows where the value is True.

In [16]:
cs_universities = df_2022_rankings[df_2022_rankings['is_computer_science']==True]

In [17]:
cs_universities

rank_order  rank                                 name        location  \
0            10     1                 University of Oxford  United Kingdom   
1            20    =2   California Institute of Technology   United States   
2            30    =2                   Harvard University   United States   
3            40     4                  Stanford University   United States   
4            50    =5              University of Cambridge  United Kingdom   
..          ...   ...                                  ...             ...   
193        1940  =193  Queensland University of Technology       Australia   
194        1950  =193                 Texas A&M University   United States   
196        1970  =197     University of Erlangen-Nuremberg         Germany   
197        1980  =197                 University of Geneva     Switzerland   
199        2000  =197          Sapienza University of Rome           Italy   

                                      subjects_offered  is_computer_science  
0    Accounting & Finance,General Engineering,Commu...                 True  
1    Languages, Literature & Linguistics,Economics ...                 True  
2    Mathematics & Statistics,Civil Engineering,Lan...                 True  
3    Physics & Astronomy,Computer Science,Politics ...                 True  
4    Business & Management,General Engineering,Art,...                 True  
..                                                 ...                  ...  
193  Languages, Literature & Linguistics,Education,...                 True  
194  Art, Performing Arts & Design,Sociology,Chemis...                 True  
196  Archaeology,Computer Science,Sport Science,Bio...                 True  
197  Politics & International Studies (incl Develop...                 True  
199  Politics & International Studies (incl Develop...                 True  

[193 rows x 6 columns]

As you can see, the number of rows is now reduced to 193. Now out of these, we want the top 10 ranked universities.

In [18]:
top_cs_universities = cs_universities[:10]

In [19]:
top_cs_universities[['name', 'rank', 'location']]

name rank        location
0                   University of Oxford    1  United Kingdom
1     California Institute of Technology   =2   United States
2                     Harvard University   =2   United States
3                    Stanford University    4   United States
4                University of Cambridge   =5  United Kingdom
5  Massachusetts Institute of Technology   =5   United States
6                   Princeton University    7   United States
7     University of California, Berkeley    8   United States
8                        Yale University    9   United States
9              The University of Chicago   10   United States

### 2. Top-10 Computer Science Universities in the US
We will use the same "Slicing" as before to get the top-10 universities. 

We also have the __is_computer_science__ column from before that we will use to get universities offering Computer Science.

We now need to get universities whose *location* is "United States"

In [20]:
cs_universities_us = cs_universities[cs_universities['location']=='United States']

In [21]:
cs_universities_us.head(5)

rank_order rank                                   name       location  \
1          20   =2     California Institute of Technology  United States   
2          30   =2                     Harvard University  United States   
3          40    4                    Stanford University  United States   
5          60   =5  Massachusetts Institute of Technology  United States   
6          70    7                   Princeton University  United States   

                                    subjects_offered  is_computer_science  
1  Languages, Literature & Linguistics,Economics ...                 True  
2  Mathematics & Statistics,Civil Engineering,Lan...                 True  
3  Physics & Astronomy,Computer Science,Politics ...                 True  
5  Mathematics & Statistics,Languages, Literature...                 True  
6  Languages, Literature & Linguistics,Biological...                 True

As you can see, now we have a dataframe containing only universities in the US. It is straightforward to extract the top-10 using the slicing method now.

In [22]:
top_cs_universities_us = cs_universities_us[:10]

In [23]:
top_cs_universities_us[['name', 'rank', 'location']]

name rank       location
1      California Institute of Technology   =2  United States
2                      Harvard University   =2  United States
3                     Stanford University    4  United States
5   Massachusetts Institute of Technology   =5  United States
6                    Princeton University    7  United States
7      University of California, Berkeley    8  United States
8                         Yale University    9  United States
9               The University of Chicago   10  United States
10                    Columbia University   11  United States
12               Johns Hopkins University  =13  United States

Let us cross-check the number of rows in the top CS universities. This should be equal to 10.

In [24]:
print(top_cs_universities_us.shape[0])

10


## What other questions can we answer?
 - In 2021, what were the top-5 universities in the United Kingdom that offered __Business & Management__?
 - What were the top-10 universities overall in the United States in 2014?
 - How many universities in the United States offered __Sociology__ in 2015?
 - What courses were offered at __University of Texas at Austin__ in 2019?